infer/environment.cc

```diff
@@ -831,6 +831,16 @@ core::TypePtr Environment::processBinding(core::Context ctx, const cfg::CFG &inW
                     for (auto &err : it->main.errors) {
                         ctx.state._error(std::move(err));
                     }
+
+                    // The method here should always exists() and return true for `isMethod()`.
+                    // However, when initializing a class (e.g. `Object.new`), that is not the case. This guards against that.
+                    if (it->main.method.exists() && it->main.method.data(ctx)->isMethod() &&
```

Comment:
Unfortunately we need to check that the method is a method here. I didn't spend a lot of time debugging, but from initial debugging it looked like `it->main.method.exists()` is false when initializing an object (e.g. `Object.new`). I think we'll need to fix this separately though and remove the check here when that's fixed.

Reply:
ok i'd be more comfortable fixing that first and landing private methods second

Reply:
That's fine. Let me take a stab a figuring out what's going on.

Reply:
@jez 
So, I've found the issue.
It's pretty easy to reproduce by removing the check for `exists()` and `isMethod()` in this line and then running sorbet on a file containing:
```
# typed: true
Object.new
```
It happens when `#initialize` is not overloaded for a class.

Tracing through the code `main.method.exists()` becomes false during the dispatch to `#initialize` from `.new`. In that call:

https://github.com/sorbet/sorbet/blob/master/core/types/calls.cc#L1173-L1181

`main` is replaced with the `main` for `#initialize`. `#initialize` is special cased in `dispatchCallSymbol`:

https://github.com/sorbet/sorbet/blob/master/core/types/calls.cc#L497-L511

That code sets the `SymbolRef` for the returned `DispatchResult` to be `Symbols::noSymbol()`, which will return `false` on `exists()`.

Now the question is how we fix that. Could we create a new static symbol ref, e.g.:
```
static SymbolRef DefaultInitializeMethod() {
  return SymbolRef(nullptr, SymbolRef::Kind::Method, XXXX); // TODO: figure out the right id that is != 0
}
```

and use that instead of `Symbols::noSymbol()` ? Since the _id would be `!= 0`, `exists()` should return `true`, and since it has `SymbolRef::Kind::Method`, `isMethod` should return `true` as well. Do you see any downside to this approach?

Reply:
Just tried this and that doesn't work. It seems to be due to sorbet not being able to lookup the `Symbol` from the `SymbolRef`. I looked into adding an initialize method as `GlobalState.cc`, but that made me think about the comment above about not being able to define it on `BasicObject` and having to special case it. Is it still the case that that is needed? Do you know what the special cases is where that would shadow the real definition? It seem like the right solution to me. But perhaps you can think of something simpler?

Reply:
The thing I’m more worried about is not that a symbol doesn’t exist (we might be able to work around some stuff with initialize)

But I’m very concerned that if the dispatch main exists that it’s not a method symbol

Reply:
@jez I don't entirely follow what you mean by your last sentence, can you expand a bit on that?

Reply:
Ah I see. Let me double check but pretty sure all the cases I've seen so far only needed the `.exists()` guard, agreed it would be worry-some if it was suddenly not a method in here.